agent/agentcontainers/api_test.go

```diff
@@ -3775,43 +3705,73 @@ func TestDevcontainerDiscovery(t *testing.T) {
 					mClock = quartz.NewMock(t)
-					mDCCLI = acmock.NewMockDevcontainerCLI(gomock.NewController(t))
+
+					upCalledMu  sync.Mutex
+					upCalledFor = map[string]bool{}
+
+					fCCLI  = &fakeContainerCLI{}
+					fDCCLI = &fakeDevcontainerCLI{
+						configMap: tt.configMap,
+						up: func(_, configPath string) (string, error) {
+							upCalledMu.Lock()
+							upCalledFor[configPath] = true
+							upCalledMu.Unlock()
+							return "", nil
+						},
+					}
 
 					r = chi.NewRouter()
 				)
 
-				// Given: We setup our mocks. These mocks handle our expectations for these
-				// tests. If there are missing/unexpected mock calls, the test will fail.
-				tt.setupMocks(mDCCLI)
-
 				api := agentcontainers.NewAPI(logger,
 					agentcontainers.WithClock(mClock),
 					agentcontainers.WithWatcher(watcher.NewNoop()),
 					agentcontainers.WithFileSystem(initFS(t, tt.fs)),
 					agentcontainers.WithManifestInfo("owner", "workspace", "parent-agent", "/home/coder"),
-					agentcontainers.WithContainerCLI(&fakeContainerCLI{}),
-					agentcontainers.WithDevcontainerCLI(mDCCLI),
+					agentcontainers.WithContainerCLI(fCCLI),
+					agentcontainers.WithDevcontainerCLI(fDCCLI),
 					agentcontainers.WithProjectDiscovery(true),
 					agentcontainers.WithDiscoveryAutostart(true),
 				)
 				api.Start()
-				defer api.Close()
 				r.Mount("/", api.Routes())
 
-				// When: All expected dev containers have been found.
+				// Given: We allow the discover routing to progress
+				var got codersdk.WorkspaceAgentListContainersResponse
 				require.Eventuallyf(t, func() bool {
 					req := httptest.NewRequest(http.MethodGet, "/", nil).WithContext(ctx)
 					rec := httptest.NewRecorder()
 					r.ServeHTTP(rec, req)
 
-					got := codersdk.WorkspaceAgentListContainersResponse{}
+					got = codersdk.WorkspaceAgentListContainersResponse{}
 					err := json.NewDecoder(rec.Body).Decode(&got)
 					require.NoError(t, err)
 
-					return len(got.Devcontainers) >= tt.expectDevcontainerCount
+					upCalledMu.Lock()
+					upCalledCount := len(upCalledFor)
+					upCalledMu.Unlock()
+
+					return len(got.Devcontainers) >= tt.expectDevcontainerCount && upCalledCount >= tt.expectUpCalledCount
 				}, testutil.WaitShort, testutil.IntervalFast, "dev containers never found")
 
-				// Then: We expect the mock infra to not fail.
+				// Close the API. We expect this not to fail because we should have finished
+				// at this point.
+				err := api.Close()
+				require.NoError(t, err)
+
+				// Then: We expect to find the expected devcontainers
+				assert.Len(t, got.Devcontainers, tt.expectDevcontainerCount)
+
+				// And: We expect `up` to have been called the expected amount of times.
+				assert.Len(t, upCalledFor, tt.expectUpCalledCount)
+
+				// And: `up` was called on the correct containers
+				for configPath, config := range tt.configMap {
+					autoStart := config.Configuration.Customizations.Coder.AutoStart
+					wasUpCalled := upCalledFor[configPath]
```

Comment:
Will this handle the case where we call `up` for something not in `config.Configuration.Customizations.Coder.AutoStart`?